File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        int answer = backTrack(numbers, target, 0);
        return answer;
    }
    
    public int backTrack(int[] numbers, int target, int depth) {
        if (depth == numbers.length) {
            int sum = 0;
            for (int n : numbers) {
                sum += n;
            }
            
            return sum == target ? 1 : 0;
        }
        
        int result = 0;
        
        result += backTrack(numbers, target, depth+1);
        numbers[depth] *= -1;
        
        result += backTrack(numbers, target, depth+1);
        numbers[depth] *= -1;
        
        return result;
    }
    
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque
def solution(numbers, target):
    answer = 0
    size = len(numbers)
    queue = deque()
    queue.append((numbers[0], 1))
    queue.append((-numbers[0], 1))
    
    while queue:
        number, depth = queue.popleft()
        depth += 1
        if depth <= size:
            queue.append((number+numbers[depth-1], depth))
            queue.append((number-numbers[depth-1], depth))
        else:
            if number == target:
                answer += 1
    
    
    return answer

